src/coreComponents/mesh/generators/VTKMeshGenerator.cpp

```diff
@@ -198,95 +151,147 @@ vtkSmartPointer< vtkUnstructuredGrid > loadVTKMesh( Path const & filePath )
   return loadedMesh;
 }
 
+/**
+ * @brief Compute assignments of mesh partitions produced by vtkRedistributeDataSetFilter to ranks.
+ * @param numParts number of partitions (cuts)
+ * @param numRanks number of MPI ranks
+ * @return a vector of rank assignments
+ *
+ * This function merges partitions starting from the back, until the desired number is reached.
+ * It is a stripped down version of vtkDIYKdTreeUtilities::ComputeAssignments(). We cannot call
+ * that function directly because it's part of a VTK private module (i.e. header not installed).
+ * We rely on copying the assignment procedure for lack of a better solution.
+ * In particular, there is no API to extract this information out of vtkRedistributeDataSetFilter.
+ */
+std::vector< int > computePartitionAssignments( int const numParts, int const numRanks )
+{
+  GEOSX_ERROR_IF_NE_MSG( numParts, vtkMath::NearestPowerOfTwo( numRanks ),
+                         "Number of partitions is not a nearest power of 2 of number of ranks."
+                         "This could indicate a change in VTK implementation of mesh partitioner." );
+
+  std::vector< int > assignments( numParts );
+  std::iota( assignments.begin(), assignments.end(), 0 );
+
+  int numToMerge = numParts - numRanks;
+  int rank = numRanks - 1;
+  for( int cc = numParts - 2; cc >= 0 && numToMerge > 0; cc -= 2, --numToMerge, --rank )
+  {
+    assignments[cc] = assignments[cc + 1] = rank;
+  }
+  return assignments;
+}
```

Comment:
This code is basically a simplified version of https://gitlab.kitware.com/vtk/vtk/-/blob/d79e2977b601db1e0592c02b1536f39b6b367314/Filters/ParallelDIY2/vtkDIYKdTreeUtilities.cxx#L511
(we don't need it to be as general because we always know that `numParts` will be a nearest power of 2 of `numRanks`).

I would've loved to use VTK's function directly, but that particular class `vtkDIYKdTreeUtilities` is considered private to the VTK module and it's header is not installed by `make install`. Replicating the code is a fairly dangerous form of dependency on implementation details. That said, I don't expect this behavior to change often, and if it does, it can only hurt us during a future VTK version upgrade, when we can detect and fix it in the same PR.

In the meantime, I will be looking for ways to get a more stable solution in place, for example by exposing the necessary information through `vtkRedistributeDataSetFilter`. I opened a [discussion](https://discourse.vtk.org/t/vtkredistributedatasetfilter-cuts-to-mpi-ranks/7927) with a person who's been actively working on this particular piece of VTK recently (judging by commit history), so hopefully we can find a permanent solution.

Reply:
```c++
std::vector< int > computePartitionAssignments( int const numParts, int const numRanks )
{
  std::vector< int > assignments;
  assignments.reserve( numParts );
  std::div_t const euclid = std::div( numParts, numRanks );
  
  for( int i = 0; i < numRanks; ++i ) {
      int const bound = i < numRanks - euclid.rem ? euclid.quot : euclid.quot + 1;
      for( int j = 0; j < bound; ++j ) {
          assignments.push_back( i );
      }
  }
  
  return assignments;
}
```
IIUC, isn't this implementation equivalent (with less problematic corner cases) and a bit more readable?

Reply:
BTW, the original code is
```cpp
assert(num_blocks == vtkMath::NearestPowerOfTwo(num_blocks));
```
which is strange (`num_blocks` on both sides of the `==`) 🤯

Reply:
> BTW, the original code is
> 
> ```c++
> assert(num_blocks == vtkMath::NearestPowerOfTwo(num_blocks));
> ```
> 
> which is strange (`num_blocks` on both sides of the `==`) exploding_head

I mean it's a valid way of checking that something is a power of 2. A strange one for sure, given that they also have a function `vtkMath::IsPowerOfTwo`.

Reply:
> IIUC, isn't this implementation equivalent (with less problematic corner cases) and a bit more readable?
> Or with simpler `int const bound = i < euclid.rem ? euclid.quot + 1: euclid.quot;` if you want to start filling from the `front` and not the `back`.

The funny part is, the original code is inconsistent about this: when `numRanks <= numParts` (normal case), it puts fewer parts on lower ranks and more parts on higher ranks; but when `numRanks > numParts` (edge case), it puts all parts on low ranks, and 0 on higher ones.

Reply:
> I mean it's a valid way of checking that something is a power of 2. A strange one for sure, given that they also have a function `vtkMath::IsPowerOfTwo`.

In your code you do compare `vtkMath::NearestPowerOfTwo( num_blocks )` with `num_parts`, not `num_blocks`. Personally, I doubt the code from `vtk` is intended...
Also, I expect this `assert` statement not to be activated with `NDEBUG`? (It's most probably the standard `assert`, not the `vtkAssert`). So one may end up calling it with all kinds of parameters?

If you test the `vtk` code with `numParts = 8` and `numRanks = 2`, you get `-2, -2, -1, -1, 0, 0, 1, 1` while the euclidian version gives `0, 0, 0, 0, 1, 1, 1, 1`. 🤷